src/compiler/checker.ts

```diff
@@ -4716,19 +4731,33 @@ namespace ts {
             return result;
         }
 
+        function cloneTypePredicate(predicate: TypePredicate, mapper: TypeMapper): ThisTypePredicate | IdentifierTypePredicate {
+            if (predicate.kind === TypePredicateKind.Identifier) {

```

Comment:
You wouldn't need to check the parameter name, you could just check `parameterIndex === -1`. I don't think the differences are as extensive as you make them out to be, the check would be no more expensive than what you have now, and I think you could eliminate a fair bit of code. Also, the AST uses a single type for the two forms, so why not do the same in the checker.